Mathlib/Order/Lattice.lean

```diff
@@ -1279,3 +1283,89 @@ end ULift
 --To avoid noncomputability poisoning from `Bool.completeBooleanAlgebra`
 instance Bool.instPartialOrder : PartialOrder Bool := inferInstance
 instance Bool.instDistribLattice : DistribLattice Bool := inferInstance
+
+/-!
+### Lattice Congruences
+-/
+
+/-- An equivalence relation is a congruence relation for the latice structure if it is compatible
+with the `inf` and `sup` operations. -/
+structure LatticeCon (α) [Lattice α] extends Setoid α where
+  inf : ∀ {w x y z}, r w x → r y z → r (w ⊓ y) (x ⊓ z)
+  sup : ∀ {w x y z}, r w x → r y z → r (w ⊔ y) (x ⊔ z)
+
+lemma r_iff_r_inf_sup [Lattice α] (c : LatticeCon α) {x y : α} :
+    c.r x y ↔ c.r (x ⊓ y) (x ⊔ y) := by
+  constructor
+  · intro h
+    exact c.trans (b := y) (by
+      conv_rhs => rw [← inf_idem y]
+      exact c.inf h (c.refl y)) (by
+        conv_lhs => rw [← sup_idem y]
+        exact c.sup (c.symm h) (c.refl y))
+  · intro h
+    exact c.trans (b := x ⊓ y) (by
+        conv_lhs => rw [← inf_sup_self (a := x) (b := y)]
+        conv_rhs => rw [← inf_idem x, inf_assoc]
+        exact c.inf (c.refl x) (c.symm h)) (by
+        conv_rhs => rw [← inf_sup_self (a := y) (b := x), inf_comm, sup_comm]
+        conv_lhs => rw [← inf_idem y, ← inf_assoc]
+        exact c.inf h (c.refl y))
+
+lemma closed_interval [Lattice α] {r : α → α → Prop}
+    (h₂ : ∀ ⦃x y : α⦄, r x y ↔ r (x ⊓ y) (x ⊔ y))
+    (h₄ : ∀ ⦃x y t : α⦄, x ≤ y → r x y → r (x ⊓ t) (y ⊓ t) ∧ r (x ⊔ t) (y ⊔ t))
+    (a b c d : α) (hab : a ≤ b) (hbd : b ≤ d) (hac : a ≤ c) (hcd : c ≤ d) (had : r a d) :
+    r b c := by
+  rw [h₂]
+  conv_lhs => rw [← inf_eq_left.mpr inf_le_sup]
+  conv_rhs => rw [← inf_eq_right.mpr (sup_le hbd hcd)]
+  apply (h₄ (inf_le_of_left_le hbd) _).1
+  rw [← sup_eq_right.mpr (le_inf hab hac), ← sup_eq_left.mpr (inf_le_of_left_le hbd)]
+  exact (h₄ (le_trans hab hbd) had).2
+
+lemma transitive [Lattice α] {r : α → α → Prop}
+    (h₂ : ∀ ⦃x y : α⦄, r x y ↔ r (x ⊓ y) (x ⊔ y))
+    (h₃ : ∀ ⦃x y z : α⦄, x ≤ y → y ≤ z → r x y → r y z → r x z)
+    (h₄ : ∀ ⦃x y t : α⦄, x ≤ y → r x y → r (x ⊓ t) (y ⊓ t) ∧ r (x ⊔ t) (y ⊔ t)) :
+    ∀ {x y z : α}, r x y → r y z → r x z := by
+  intro x y z hxy hyz
+  exact closed_interval h₂ h₄ (x ⊓ y ⊓ z) _ _ (x ⊔ y ⊔ z) (by rw [inf_assoc]; exact inf_le_left)
+    (by rw [sup_assoc]; exact le_sup_left) inf_le_right le_sup_right
+    (h₃ (by rw [inf_assoc]; exact inf_le_of_right_le inf_le_sup)
+    (by rw [sup_assoc]; exact le_sup_right) (h₃
+    (by rw [inf_assoc]; exact inf_le_right) inf_le_sup (by
+      conv_lhs => rw [inf_comm x, inf_assoc, inf_inf_distrib_left, inf_comm _ x]
+      conv_rhs => rw [← inf_eq_right.mpr (le_trans inf_le_left le_sup_right)]
+      exact (h₄ inf_le_sup (h₂.mp hxy)).1) (h₂.mp hyz)) (by
+      conv_rhs => rw [sup_comm x, sup_assoc, sup_sup_distrib_left, sup_comm _ x]
+      conv_lhs => rw [← sup_eq_right.mpr (le_trans inf_le_right le_sup_left)]
+      exact (h₄ inf_le_sup (h₂.mp hxy)).2))
```

Comment:
Will you ever use these lemmas besides proving `LatticeCon.mk3`? If not, I would suggest making them private

Reply:
I doubt it, so I've marked them as private.